### ISSUE TITLE
Fix property interaction algorithms to correctly handle Forms, contentType and DataSchema

### ISSUE THREAD

**Issue by zolkis** [w3c/wot-scripting-api#201]
As raised in #198 and then discussed in many calls, the current interaction algorithms don't handle correctly the WoT interactions.

A WoT interaction might contain multiple [Forms](https://w3c.github.io/wot-thing-description/#form), each of which define a `contentType` and for some content types, may define `DataSchema` in the Property definition itself. 

In most cases, DataSchema will be defined, but e.g. for media types the value will be binary and the media type needs to be passed to the application (in the Form) in order to be able to parse it. But currently there is no mechanism for that.

The `readProperty()` algorithm needs to choose one of the Forms, first handle `contentType`, then if the case, handle `DataSchema` and then validate/handle the binary `value` according to these.

This issue replaces #198 as it is wider.

**Comment by zolkis** [comment]
Implementations could map WoT property read interactions to a JS interface of obtaining the data (based on [Web Platform](https://en.wikipedia.org/wiki/Web_platform) idioms). An attempt has been described [here](https://github.com/w3c/wot-scripting-api/issues/198#issuecomment-547589233):

> The proper way would be a Scripting-defined Value dictionary like the one above (and below): currently we need those 4 properties in the Web Platform to represent binary or textual data with encoding and language.
So this is my proposal to handle values in WoT Scripting:

```javascript
interface Value {
  readonly attribute USVString? mediaType;
  readonly attribute USVString? encoding;
  readonly attribute USVString? lang;
  readonly attribute DataView? data;
}
```

> Then we can create/read text, JSON or binary by choosing the encoder/decoder.

**Comment by egekorkan** [comment]
We can also look at the [fetch API for the browsers.](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)  There, one can do `response.json()` in order to obtain the JSON value of the response. Maybe we can do something similar where `readproperty` returns a value and the script programmer has to parse the data out of it like implied in the previous comment(s).

Another way I can think of this would be to give `readProperty` method an optional argument, where the programmer can specify the desired contentType. The Scripting runtime would then try to convert the payload if it can, otherwise return the binary. If this argument is not given, the runtime tries to convert to JSON by default.

I think it is also worth thinking the similar case for other Interaction Affordances, i.e. choosing an Action Form based on the content type or the Event form to subscribe to.

**Comment by zolkis** [comment]
> We can also look at the fetch API for the browsers. There, one can do response.json() in order to obtain the JSON value of the response. Maybe we can do something similar where readproperty returns a value and the script programmer has to parse the data out of it like implied in the previous comment(s).

That is right. Actually we used to have `.json()`, `.text()` etc in Web NFC as well, but after playing around for a while, we ended up to  discard them in the favor of `DataView` for simplicity. The code samples were significantly different in complexity or length, and then simplicity prevailed. See some [examples](https://w3c.github.io/web-nfc/#examples) on how to handle e.g. text, JSON and jpg with them. Also available in [Node.js](https://github.com/nodejs/node-addon-api/blob/master/doc/dataview.md). 

That said, we can still decide to have helper functions for fetching `text()`, `json()` or others.

**Comment by zolkis** [comment]
Speaking about `fetch()` API, actually that is _the_ universal API to handle WoT read interactions.
The advantage of the current Scripting API approach for reading/writing Properties is that protocol bindings are automatically handled by the implementations. Which is maybe not desired in every case.

But if Scripting API the cannot do this in a simple way (for instance, how apps can choose which Form is applied and then how to fetch values), then apps could revert to use the `fetch()` API directly, for sure.

**Comment by zolkis** [comment]
> The readProperty() algorithm needs to choose one of the Forms, first handle contentType, then if the case, handle DataSchema and then validate/handle the binary value according to these.

For the record, one proposal on the algorithm is [here](https://github.com/w3c/wot-thing-description/issues/839#issuecomment-552438034) and [here](https://github.com/w3c/wot-thing-description/issues/839#issuecomment-562210278).

**Comment by danielpeintner** [comment]
related to TD https://github.com/w3c/wot-thing-description/pull/869

**Comment by zolkis** [comment]
A relevant [comment](https://github.com/w3c/wot-thing-description/issues/875#issuecomment-579128343) from @danielpeintner 

One more reason to go with a value definition for Scripting like the one [above](https://github.com/w3c/wot-scripting-api/issues/201#issuecomment-573702999).

In the Scripting call on 30.March 2020 @danielpeintner proposed removing the `dataSchema` part from the `Value` structure. Normally it is not needed since the DataSchema can be fetched from the Thing Description.

However, in the cases when a completely different value is returned, like described by @egekorkan  [here](https://github.com/w3c/wot-thing-description/issues/875#issuecomment-579158814), it would be still needed by applications.

So the rule would be that if a DataSchema is returned with a value, it means a departure from the DataSchema defined in the TD for the Property.

**Comment by zolkis** [comment]
In addition, we could also define WotValue as a union type of `null`, `boolean` or `Value` (which handles strings + encoding + lang, various number formats, arrays and objects). Since `data` can be `null`, the only simple type we can separately represent is `boolean`, not much gain. 
Therefore going with the Value definition as above (except the name - a better name needs to be agreed).

**Comment by zolkis** [comment]
For the record, on the Scripting call today (27.4.2020) we agreed to keep this issue open even after #209 is merged, in order to experiment with options.

1. One of them is outlined [here](https://github.com/w3c/wot-scripting-api/pull/209#issuecomment-619376479).
Pro: That would be backwards compatible with the current API, only adds a new way to represent payloads that don't have a DataSchema nor a contentType that could be converted into a type supported by JSON Schema. Also on writes, an app could use any simple type, or InteractionData, or otherwise an object.
- Problem: how to expose strings (vs encoding and language: as string or as InteractionData).
One solution is to say that implementations should try to infer that from the TD (`@context`), or the scripting environment, or use default English (TBD).
We could also add accepted language and encoding options (app hints) to InteractionOptions, where an app could tell implementations its preferences. 
Then strings can be always exposed as strings, even when apps don't use these options.

2. Another option would be to always use InteractionOptions, but include a property for a parsed `value` that would mean the same thing as `any` now, i.e. simple types/types supported by JSON Schema and will be used every time the implementation can figure out the payload type and convert it to JavaScript type. For others (e.g. image contentType), it will set `value` to `null` and will set `data` to the raw bytes. Applications can then attempt to parse the payload themselves.
- Problem: on writes the same structure is used. It is slightly more inconvenient to always pass values in a dictionary rather than directly as a parameter.

```javascript
interface InteractionData {
  readonly attribute any value;  // parsed whenever possible or null otherwise
  readonly attribute DataView? data;  // raw payload bytes
  readonly attribute USVString? encoding;
  readonly attribute USVString? lang;
  readonly attribute USVString? mediaType;
  readonly attribute DataSchema? dataSchema;
}
```

**Comment by danielpeintner** [comment]
FYI: One problem I see is representing the value `null` in `value` (without TD/dataSchema knowledge)

**Comment by zolkis** [comment]
`any` covers `null` as well.

**Comment by danielpeintner** [comment]
Overall I like option 2 better.

**Comment by zolkis** [comment]
Option 2 would be preferred more by browser people, AFAICT. At least for reading.

However, writing looks bad... I would not exclude asymmetric API, for instance Web NFC is like that, infers the type from JS when writing and exposed a structure like InteractionData for reading. It's a very similar situation like we have. 

That would be option 3: use InteractionData for reading and `any` for writing. Or instead of `any`, we could use an explicit union type (but because of Web IDL peculiarities, we will end up with `any`).

**Comment by zolkis** [comment]
In the Scripting call on May 4 2020 different API style option have been discussed.
The first 2 (see next) are the same as in the comment [above](https://github.com/w3c/wot-scripting-api/issues/201#issuecomment-620182827):

1. reads resolve with `any`, writes take `any` (the current PR version)
Advantages: 
- doesn't break the API, just adds `InteractionData` for the corner cases
- developer friendly and also symmetric on reads vs writes
Disadvantages: 
- some issues with string handling on reads, vs language/encoding
- need to figure out type from value like explained [here](https://github.com/w3c/wot-scripting-api/pull/209#issuecomment-619376479) (apps on reads, implementation on writes).

2. reads resolve with `InteractionData`, writes take `InteractionData`
Advantage: same data interface for read and write
Disadvantage: 
- not developer friendly in stock version (conversion to/from binary buffer in the app), 
- adding a property `any value` is same complexity as in option 1, just that now the syntax is inconvenient for writes.

3. reads resolve with `InteractionData` (including `any value` for convenience), writes take `any` (as the `any value` on option 2 above), but parts of what is now `InteractionData` move into `InteractionOptions`
Advantages: 
- app code looks developer friendly
- solves the string problem of 1
Disadvantages:
- need for separate `WriteInteractionOptions` that extends `InteractionOptions` with Scripting-specific options
- asymmetric API.

4. reads resolve with `InteractionData` (including `any value` for convenience), writes take `any` (as in option 1 above), `InteractionData` remains the same.
Advantages:
- solves the string problem of 1
- writes are developer-friendly
Disadvantages:
- reads are less developer friendly than 1, with same implementation complexity
- asymmetric API
- doesn't solve all problems of 1.

The least resistance path has been option 1.
The other "clean" solution would be option 2, but less developer friendly, with same complexity.

**Comment by zolkis** [comment]
It has been raised on the Scripting call on May 4 2020,  whether using `any` will inflict implementation problems.
At least in the [Web API Design Cookbook](https://www.w3.org/TR/api-design/#attribute-constant-or-exception-fields) the use of `any` is recommended.
On the other hand, in this [issue discussion](https://github.com/w3c/web-nfc/issues/232) the use of `any` and `object` are not recommended for security reasons and V8 issues.
Also, data should not be exposed as attribute, but with fetcher methods.

Which leads back to the option to model the API on [`fetch()`](https://fetch.spec.whatwg.org/#fetch-api) as suggested [above](https://github.com/w3c/wot-scripting-api/issues/201#issuecomment-573717804).

That is, turn `InteractionData` into an interface with fetcher methods such in [Body](https://fetch.spec.whatwg.org/#body), e.g. `text()`, `json()`, etc. 
Parsing floating point e.g. could be done by `parseFloat(data.text())`.
Or just expose data as `DataView` and let the application handle it.

In general, the best API would be `fetch()`, since it works with URI variables, can handle streams etc on return, etc. The problem then becomes translating between protocol bindings and `fetch()`.

This would be a next step in how to expose interaction data.

**Comment by danielpeintner** [comment]
> In general, the best API would be `fetch()`, since it works with URI variables, can handle streams etc on return, etc. The problem then becomes translating between protocol bindings and `fetch()`.

Besides bindings we also need the counterpart such as *push*. Am I wrong?
Hence I don't think fetch is the best API.

**Comment by zolkis** [comment]
We can do POST with the Fetch API (in general, [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) and [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response).

Fetch is pretty complex, what I meant was a convenient WoT-specific convenience-wrapper on top, mainly around the [Body mixin](https://fetch.spec.whatwg.org/#body-mixin), see the [API](https://developer.mozilla.org/en-US/docs/Web/API/Body).

Of course there is the [Node.js fetch](https://www.npmjs.com/package/node-fetch).

**Comment by zolkis** [comment]
Inspired from the [Body mixin](https://fetch.spec.whatwg.org/#body-mixin), for reads we could always use `InteractionData` augmented (for convenience) with `json()`, eventually also `text()`. It would be better defined than the current `any`.

Playing with that a bit didn't seem to be better than to provide the `value` property [above](https://github.com/w3c/wot-scripting-api/issues/201#issuecomment-620182827) (which we should probably turn into a method `value()`). 
Examples for reads would look almost the same:

```javascript
try {
  let res = await fetch("https://tds.mythings.org/sensor11");
  let td = res.json();
  let thing = new ConsumedThing(td);
  console.log("Thing " + thing.getThingDescription().title + " consumed.");
} catch(e) {
  console.log("TD fetch error: " + e.message); },
};

let response = await thing.readProperty(“temperature”));
let value = response.value();  // implementation does the magic of parsing
if (value === undefined) {  // magic failed, app should parse it manually from bytes
     //  use a special decoder on `response.data`
     value = parseContent(response.data, 
                     thing.getThingDescription.forms[0].contentType);
}
console.log(“Temperature: “ + value);
```

For writes, I don't have any problems keeping the `any` interface convenience.
However, if we use `InteractionDataInit` there, it could handle the language and encoding hints we [discussed](#213 ) adding to `InteractionOptions`.

```javascript

dictionary InteractionDataInit {
  required any value;
  USVString? encoding;
  USVString? lang;
  USVString? mediaType;
};

[SecureContext, Exposed=Window]
interface InteractionData {
  constructor (InteractionDataInit init);

  any value();  // parsed whenever possible or undefined otherwise
  DataSchema? dataschema();

  readonly attribute DataView? data;  // raw payload bytes
  readonly attribute USVString? encoding;
  readonly attribute USVString? lang;
  readonly attribute USVString? mediaType;
};
```
The show-stopper on that option is that dictionaries should not have properties with `any` or `object` (for security reasons and due to V8 internals).

So we'd better stick with the asymmetric API IMHO so `writeProperty()` would take an `any` parameter like now.

**Comment by relu91** [comment]
In my opinion, is starting to be a nice reach interface that can work also with other protocol bindings not only HTTP. I'd like to experiment a little bit with it when it is ready inside `node-wot`

One comment about `InteractionData`. What if we change the type of `data` attribute to [ReadableStream](https://fetch.spec.whatwg.org/#concept-readablestream)? That way, we could handle also streaming data. It does not seem a really drastic change, because you can still read Buffers or Typed Arrays trough a streaming interface. This would imply that `value` function should become an async function, again really similar to [Body mixin](https://fetch.spec.whatwg.org/#body-mixin) but we some WoT sauce on it.

What do you think?

**Comment by zolkis** [comment]
Hmm, I thought about that a few times, but ReadableStream is not the same as Readable in Node.js (and node-fetch is based on the latter). Which would need different code in Node.js and in the browser.

Also, it has strict rules about consuming a stream and about readers. Not sure if we want to track all that in this API. However, handling streams would be desirable - it would just complicate the simple use cases too much.

I am open to suggestions :).

**Comment by zolkis** [comment]
Actually there is [ReadableStream polyfill]() for Node.js, so it's fine.

Now `InteractionData` would become almost the same as the [Body mixin](https://fetch.spec.whatwg.org/#body-mixin)...

Conversely, I guess we'd want to support [WritableStream](https://streams.spec.whatwg.org/#ws) for write ops?

**Comment by relu91** [comment]
> Actually there is ReadableStream polyfill for Node.js, so it's fine.

I was about to post the same (it seems that link you used to send back to this issue, I guess you mean [this](web-streams-polyfill)?).

> However, handling streams would be desirable - it would just complicate the simple use cases too much.

Can you please make an example? for the simple use cases, we still have the `value()` function. The only complication that I foresee is that it might require more work than reading from a `DataView`. We could add a function for retrieving it... but yes we are quite close to reproducing the entire `Body mixin`.  

>Conversely, I guess we'd want to support WritableStream for write ops?

Make sense.

**Comment by zolkis** [comment]
In the end I went with `ReadableStream` both for reading and writing, since as `BodyInit` includes `ReadableStream`, it's possible to construct a `Body`/`Response` from it.

The  `value` was represented as read-only attribute, not function, for using the same `Promise` as for the interaction in order to attempt the parsing. If parsing is not successful, a `ReadableStream` is presented in `InteractionData` (otherwise it _could_ be set to `null`, or provided anyway - the current text requires the latter, which is not a big deal since we already have the stream anyway from the `Body` of a `Response`.

(Here I wonder should we go into more details in the spec, down to how to use the Fetch standard to construct responses/use protocol bindings - at the moment it's behind the "request the underlying platform to..." euphemism :) ).

In addition, the `Form` object selected for the interaction is also linked in `InteractionData` and it provides content type and encoding. Since the TD should provide language, I removed the `mediaType` and `encoding` and `lang` attributes.

I think now we have a pretty neat and versatile `InteractionData` interface. My concern is about `value`, whether is it an idiomatic enough representation of types described by DataSchema (JSON schema).

I have updated the ConsumedThing example with a reader that returns `value` if it's not `undefined`, otherwise reads the stream (slightly incomplete on that end, but gives an idea).

More examples are welcome, if you'd experiment with the proposal in #209 anyway.

**Comment by relu91** [comment]
> I think now we have a pretty neat and versatile InteractionData interface
I believe that too! Just one objection about this:

> The value was represented as read-only attribute, not function, for using the same Promise as for the interaction in order to attempt the parsing. If parsing is not successful, a ReadableStream is presented in InteractionData (otherwise it could be set to null, or provided anyway - the current text requires the latter, which is not a big deal since we already have the stream anyway from the Body of a Response.

In my experience, the process of parsing data will consume the underlying stream so that it will be unusable thereafter. Therefore, protocol binding implementations should buffer all the data and then expose it as a stream if the parsing failed. This probably nulls every advantage of using streamings instead of DataViews.  
Consequently, I propose to define `value` as an asynchronous function and always return the `ReadableStream`. Applications may ignore completely the stream and just use the `value` function. This means that they are dispatching the data deserialization to the underlying runtime. On the other hand, they could use `ReadableStream` if they want more control. The only problem that I see here is that the applications will need a way to choose between `value` and `ReadableStream`. It could be the presence of a DataSchema inside an Interaction Affordance. For example:

```javascript
	/*
	* Temperature Range affordance form:
	* {...
	* "temperatureRange": 
	*     { 
	*         "description" : "Shows the operational range of this device", 
	*         "type": "object",
	*          "properties": {
	*             "min": { "type": "number" },
	*             "max": { "type": "number" }
	*          }, 
	*         "forms": [
	*             { 
	*                 "op": "readproperty", 
	*                 "href": "http://mylamp.example.com/status", 
	*              }
	*         ] 
	*     }
	*... }
	*/
	assert(td.properties.temperatureRangere.type!= undefined)
	const response = await thing.readProperty(“temperatureRange”));
	const range = await response.value()
	// range {min: 15, max: 20} 
```

Anyway, I have started working on use cases and examples. You can find a very early draft in [this gist](https://gist.github.com/relu91/05928793fbed8e95b1db0035c9038bf7).

**Comment by zolkis** [comment]
> the process of parsing data will consume the underlying stream so that it will be unusable thereafter.

True, which is why `Body` has a `bodyUsed` property.

> Consequently, I propose to define value as an asynchronous function 

I considered that, too, and it also blends better with the `Body` functions that are like 
`Promise<any> fun()`

However, I also considered an optimization:
- since the interaction already returns a `Promise` and
- the implementation knows if a `DataSchema` is defined for the interaction,
the implementation could right away parse the value and provide it as a readonly property, in which 
(positive) case the stream is consumed and needs not to be exposed, i.e. `data` could be `undefined`.
If there was no `DataSchema` or the parsing failed for other reason, then `form` is needed (for content type and encoding) and `data` could be exposed as a stream, while `value` would be `undefined`.

So applications could do what the example in the PR: if `value` is not `undefined`, use that, otherwise use the `data` stream.

IIUC, you suggest exposing both a `value()` function AND the `data` stream, which will be consumed either by 
- invoking `value()` by the app (what happens with `data` after the call resolves? what happens if the app reads from a consumed stream?), 
- or by consuming `data`. 
Should we also add a `consumed` boolean property in `InteractionData`? 
So apps could try to use `value()`, and if that rejects, they could use `data`, but if it resolves we need to define what happens when `data` is accessed.

I think this solution is more idiomatic, while the previous one is simpler (saves one cycle in the event loop and doesn't need a `consumed` flag) and it's more straightforward to use in code, but YMMV.
Let's see which would work out better.

**Comment by relu91** [comment]
I agree that the runtime could act proactively and parse right away the data. However, there might be cases when the application prefers a stream interaction over a "just give me data" mode. I gave an example [here](https://gist.github.com/relu91/05928793fbed8e95b1db0035c9038bf7#low-level-interactions-draft). 

> IIUC, you suggest exposing both a value() function AND the data stream

exactly

>  what happens with data after the call resolves? what happens if the app reads from a consumed stream?

Good points. I did an experiment using the fetch function on my Chrome console. It seems that after the body is read every other consequent call to the other reader methods fails. Basically, You can only read once 😄 . 

> Should we also add a consumed boolean property in InteractionData?

I agree, we should provide also this property.

> saves one cycle in the event loop and doesn't need a consumed flag

Yes unfortunately, we have to add the flag but I do not think that we waste a cycle in the event loop... Implementations would use that call anyway because of the need to read data from the network which is intrinsically one or more call inside the event loop. With the proposal we are just deferring the moment when the parsing is happening.  We can discuss more this during today's call!

**Comment by zolkis** [comment]
On today's Scripting call @relu91's proposal was accepted in the following form:
- `Promise<any> value()` as async function resolving with the type described DataSchema, or if that does not exist, then it will reject;
- meaning that the app could try to read the data with one of the following lower level methods:
    1. `Promise<ArrayBuffer> arrayBuffer();` (coming from the `Body` mixin)
    2. check the new flag `consumed` or `streamUsed` or `dataUsed` (modeled after `bodyUsed` in the `Body` mixing). If `false`, then the app could try reading the exposed `stream` of type `ReadableStream`. Examples: image encoded as base64.

**Question 1**: which name to use for the flag? I suggest `dataUsed` since refers to `data` and is homomorphic with `bodyUsed`.

The resulting `InteractionData` interface becomes very close to the `Body` mixin. Essentially `value()` replaces `text()` and `json()`, we don't support `blob()` and we get the (WoT-specific) `Form` used for the WoT interaction in a different way.

**Question 2**. 
Instead of the `Form` object itself, we could expose the `formIndex` like with `InteractionOptions`, though it's less intuitive and exposing the object is free, convenient, and more idiomatic. 
So I'd use the object.

**Question 3**.
In addition, we return the DataSchema as well (though that could be obtained from the TD as well as `thing.getThingDescription().properties[propname]`, so I am thinking we could dismiss it.

**Question 4**.
Corner case: what to do if there is no DataSchema, but the Form's `contentType` is supported by DataSchema, such as `text/plain` or `application/json`? Should the implementation resolve `value()` with the parsed type? I'd argue yes, since if in the rare case the app really wants to handle the stream, it would likely NOT use `value()` to start with, but read the stream right away.

**Comment by zolkis** [comment]
On Question 3 above, while rewriting the reading/parsing algorithms it seems much easier to maintain the DataSchema at the InteractionData level, too (either as an internal slot, or as a property). It's free and convenient, so for the moment I am keeping it around.

**Comment by relu91** [comment]
> which name to use for the flag? I suggest dataUsed since refers to data and is homomorphic with bodyUsed.

thumb up for `dataUsed`

> Instead of the Form object itself, we could expose the formIndex like with InteractionOptions, though it's less intuitive and exposing the object is free, convenient, and more idiomatic.
So I'd use the object.

Let's go for the Form object it saves some LOC to retrieve back the form from a TD.

> In addition, we return the DataSchema as well (though that could be obtained from the TD as well as thing.getThingDescription().properties[propname], so I am thinking we could dismiss it.

> On Question 3 above, while rewriting the reading/parsing algorithms it seems much easier to maintain the DataSchema at the InteractionData level, too (either as an internal slot, or as a property). It's free and convenient, so for the moment I am keeping it around.

is there a use case where it is useful to have DataSchema after the value is parsed? Except for this doubt I am fine with keeping DataSchema around. 

> Corner case: what to do if there is no DataSchema, but the Form's contentType is supported by DataSchema, such as text/plain or application/json? Should the implementation resolve value() with the parsed type? I'd argue yes, since if in the rare case the app really wants to handle the stream, it would likely NOT use value() to start with, but read the stream right away

Good point! just a couple of questions:  What would it be the return type of text/plain? a string? What about application/xml? Is it possible to map XML to JSON without a schema?

**Comment by zolkis** [comment]
> is there a use case where it is useful to have DataSchema after the value is parsed? 

Mainly convenience and clarity. The information (basically the interaction description) is also available from the TD via `getThingDescription().properties[<interaction-name>]`. 
This just makes it one de-reference closer and plays well with the fact we also include the `Form` used for the interaction.

> Corner case: what to do if there is no DataSchema

In the last [commit](https://github.com/w3c/wot-scripting-api/pull/209/commits/49c338bc1c6e8421e07f18b7ca9cc882dc7daaa7), actually I have used the simpler choice, just expose as a stream in that case, because of the complexity involved otherwise. We can figure out and modify that later, but I think this is fairly good and in line with the TD spec.

Also, beware that because of this choice and the [related text in the TD spec](https://www.w3.org/TR/wot-thing-description/#sec-data-schema-vocabulary-definition), I made the assumption in the reading algorithm that a DataSchema is always associated to content types that are either `*json*` or there is a mapping available to JSON schema that the implementation could use. If this assumption is not true, we should modify the TD spec and add more clarity on this. 

As a consequence, I basically used the steps of the `json()` function from `Body` (the Fetch Standard) to parse the bytes and then check the DataSchema of the interaction for the imposed constraints on the values. This should cover all DataSchema subclasses AFAICT - please double check the algorithms especially [from this section](https://pr-preview.s3.amazonaws.com/zolkis/wot-scripting-api/pull/209.html#the-interactiondata-interface).

**Comment by danielpeintner** [comment]
FYI: I [pushed](https://github.com/w3c/wot-scripting-api/commit/b1ef74446342069d7789814e3ff2419fc6165514) a snapshot version to [npm](https://www.npmjs.com/package/wot-typescript-definitions) with the latest changes so that everyone can easily test it out.

A minor question: why is `dataUsed` not optional while `data` itself is optional. I do see that [fetch Body](https://fetch.spec.whatwg.org/#body) does the same but it is not clear to me either.

**Comment by zolkis** [comment]
Based on playing with the examples, it would be nice to be "safe" to call `value()` say in a `console.log()`, i.e. if it would be synchronous and if it would return `undefined` on error. 

Also, calling `value()` should not preempt the option of using the `data` stream if possible. Constrained implementations could do that, and set `dataUsed`, but in general if possible, I'd expose the stream anyway. Also, should apps in constrained implementations make another network call to fetch value because the impl consumed the stream?

**Comment by relu91** [comment]
> Based on playing with the examples, it would be nice to be "safe" to call value() say in a console.log(), i.e. if it would be synchronous and if it would return undefined on error.

Nowadays, synchronous is just one `await` away. I wouldn't border too much, but I might be influenced by my personal programming style. 

> Also, calling value() should not preempt the option of using the data stream if possible. Constrained implementations could do that, and set dataUsed, but in general if possible, I'd expose the stream anyway.

See [my comment]( https://github.com/w3c/wot-scripting-api/pull/209/files/6f55fdebcd069e18b153712dec7cd93855596446#r431849309 ). Anyway, it is a relaxing of the specs that we could give, but I do not see the real reason. 

>  Also, should apps in constrained implementations make another network call to fetch value because the impl consumed the stream?

In my view, using `value` function is an agreement. The script agrees to trust the runtime for parsing if something goes wrong there are two possibilities:
- I am miss using the function (i.e. no data schema, content-type no json). Then the stream should be readable because the runtime does not need to read the data. 
- The data was malformed/network error/wrong data schema, in such cases, everything else should fail.

**Comment by zolkis** [comment]
OK, let's try to go with figuring out all possible reasons why `value()` could fail, _before_ it would disturb the `data` stream. 

BTW it would be cheap to have an internal slot with the stream _source_, out of which multiple `ReadableStream` objects could be created and the impl would manage the data ownership and lifecycle of the source.

**Comment by zolkis** [comment]
Fixed by #209.
